### PR TITLE
[SPARK-15826] [CORE] PipedRDD to allow configurable char encoding

### PR DESCRIPTION
**Pull request by tejasapatil**
## What changes were proposed in this pull request?

Link to jira which describes the problem: https://issues.apache.org/jira/browse/SPARK-15826

The fix in this PR is to allow users specify encoding in the pipe() operation. For backward compatibility, 
keeping the default value to be system default.
## How was this patch tested?

Ran existing unit tests

### CODE REVIEW

**Comment by tejasapatil** [comment]
ok to test

File: core/src/main/scala/org/apache/spark/rdd/PipedRDD.scala [hunk 1]
```diff
@@ -129,7 +130,7 @@ private[spark] class PipedRDD[T: ClassTag](
       override def run(): Unit = {
         val err = proc.getErrorStream
         try {
-          for (line <- Source.fromInputStream(err).getLines) {
+          for (line <- Source.fromInputStream(err)(StandardCharsets.UTF_8).getLines) {
```
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/rdd/PipedRDD.scala]: nit: use `scala.io.Codec.UTF8` for Scala source

File: core/src/main/scala/org/apache/spark/rdd/PipedRDD.scala [hunk 2]
```diff
@@ -171,7 +172,7 @@ private[spark] class PipedRDD[T: ClassTag](
     }.start()
 
     // Return an iterator that read lines from the process's stdout
-    val lines = Source.fromInputStream(proc.getInputStream).getLines()
+    val lines = Source.fromInputStream(proc.getInputStream)(StandardCharsets.UTF_8).getLines
```
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/rdd/PipedRDD.scala]: nit: use scala.io.Codec.UTF8 for Scala source

File: core/src/main/scala/org/apache/spark/rdd/PipedRDD.scala [hunk 3]
```diff
@@ -147,7 +148,7 @@ private[spark] class PipedRDD[T: ClassTag](
       override def run(): Unit = {
         TaskContext.setTaskContext(context)
         val out = new PrintWriter(new BufferedWriter(
-          new OutputStreamWriter(proc.getOutputStream), bufferSize))
+          new OutputStreamWriter(proc.getOutputStream, StandardCharsets.UTF_8), bufferSize))
```
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/rdd/PipedRDD.scala]: This is the std out from another process. It should use the system default encoding, otherwise it may not work.

**Comment by zsxwing** [comment]
@tejasapatil what if the command uses non-UTF8 encoding? I feel the correct way is allowing the user to set the encoding.

**Comment by tejasapatil** [comment]
@zsxwing : I do like the idea of making this completely configurable by user. Are you ok with adding a new param to the `pipe()` function ? Another option could be to create a new spark config but I think it will be a bad idea to add more configurations.

File: core/src/main/scala/org/apache/spark/rdd/PipedRDD.scala (hunk 1 above)
Comment (srowen) [inline core/src/main/scala/org/apache/spark/rdd/PipedRDD.scala]: I think we've been using StandardCharsets pretty consistently for the constant (?). Or should.
It's a fair point about how to read the input from a local system process, though I still kinda imagine that UTF-8 is the safer default. It's nice to not make the behavior dependent on the local env. I think a well-behaved program will consistently output one encoding and that encoding is UTF-8. That is, just because the local env is Shift_JIS doesn't really suggest that apps on the machine are using Shift_JIS by default.

File: core/src/main/scala/org/apache/spark/rdd/PipedRDD.scala (hunk 2 above)
Comment (tejasapatil) [inline core/src/main/scala/org/apache/spark/rdd/PipedRDD.scala]: done

File: core/src/main/scala/org/apache/spark/rdd/PipedRDD.scala (hunk 3 above)
Comment (tejasapatil) [inline core/src/main/scala/org/apache/spark/rdd/PipedRDD.scala]: done

**Comment by sadikovi** [comment]
@tejasapatil Can you also hard-code UTF-8 for `stderr`?

File: core/src/main/scala/org/apache/spark/rdd/PipedRDD.scala (hunk 1 above)
Comment (tejasapatil) [inline core/src/main/scala/org/apache/spark/rdd/PipedRDD.scala]: @zsxwing : I don't agree that using system default encoding would be a good idea. One can see different result on different systems which would be annoying. A user can run a binary in pipe() operator and let its stderr stream emit data with a character encoding that he/she wishes to.

@srowen : I agree that UTF-8 is a safer default. Making this configurable would be the ideal thing to do.

**Comment by tejasapatil** [comment]
@sadikovi : That will not work in all cases. See : https://github.com/apache/spark/pull/13563#discussion_r66318006

**Comment by srowen** [comment]
Why would stderr be different from stdout?

File: core/src/main/scala/org/apache/spark/rdd/PipedRDD.scala (hunk 1 above)
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/rdd/PipedRDD.scala]: I misunderstood the codes here. Yep, stderr should use `UTF-8` too.

**Comment by sadikovi** [comment]
Can you give an example where it would not work? I was actually thinking about adding codec for stderr with malformed rule of CodingErrorAction.REPLACE to make it safer.

**Comment by zsxwing** [comment]
@rxin what do you think about adding an encoding parameter to `PipedRDD`? I'm inclined to it since it will resolve this issue essentially.

**Comment by rxin** [comment]
Are there cases when we need non UTF-8 encoding?

**Comment by zsxwing** [comment]
> Are there cases when we need non UTF-8 encoding?

It depends on the encoding of the external commands. Some commands may just use the default system encoding. If we hard code UTF-8, we will break those cases.

**Comment by rxin** [comment]
Got it - yea it'd make sense to make this configurable.

**Comment by tejasapatil** [comment]
@rxin @zsxwing Thanks for confirming. I will work on the change. Just to clarify, I will use the configuration value for interactions with all the streams (stderr,  stdout and stdin).

**Comment by srowen** [comment]
Yes that was the original argument for not hard-coding UTF-8 here. I'd more readily believe that a process uses UTF-8 regardless of the platform encoding. But, making it configurable and defaulting to UTF-8 sounds OK to me FWIW

**Comment by zsxwing** [comment]
> I'd more readily believe that a process uses UTF-8 regardless of the platform encoding.

This is about stdout and stderr. I think most processes won't hard code UTF-8 for them. Otherwise, it may not work when the user just runs such processes in a non-UTF8 shell.

**Comment by tejasapatil** [comment]
ok to test

File: core/src/main/scala/org/apache/spark/rdd/RDD.scala [hunk 4]
```diff
@@ -734,12 +737,14 @@ abstract class RDD[T: ClassTag](
       printPipeContext: (String => Unit) => Unit = null,
       printRDDElement: (T, String => Unit) => Unit = null,
       separateWorkingDir: Boolean = false,
-      bufferSize: Int = 8192): RDD[String] = withScope {
+      bufferSize: Int = 8192,
+      encoding: Charset = StandardCharsets.UTF_8): RDD[String] = withScope {
```
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/rdd/RDD.scala]: Use `Codec` for Scala API.

File: core/src/main/scala/org/apache/spark/rdd/PipedRDD.scala [hunk 5]
```diff
@@ -47,7 +48,8 @@ private[spark] class PipedRDD[T: ClassTag](
     printPipeContext: (String => Unit) => Unit,
     printRDDElement: (T, String => Unit) => Unit,
     separateWorkingDir: Boolean,
-    bufferSize: Int)
+    bufferSize: Int,
+    encoding: Charset)
```
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/rdd/PipedRDD.scala]: Use `Codec` for Scala API.

File: core/src/main/scala/org/apache/spark/rdd/RDD.scala (hunk 4 above)
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/rdd/RDD.scala]: I suggest using `Codec.defaultCharsetCodec` as the default value, so that existing applications won't be broken.

Comment (tejasapatil) [inline core/src/main/scala/org/apache/spark/rdd/RDD.scala]: @zsxwing 

>> Use Codec for Scala API.

`Codec` is scala specific. I intentionally did not use that because I wanted Java and Scala APIs to accept the same param types.

Anyways, based on the test case failures, none of Charset or Codec would work because they need to be serializable (I see `org.apache.spark.SparkException: Task not serializable` while running this change). I will use `String` instead. Is that fine ?

>> I suggest using Codec.defaultCharsetCodec as the default value

Thanks for catching that. Its unfortunate to not have UTF8 as the default ... but backward compatibility is far more important.

```
Caused by: java.io.NotSerializableException: sun.nio.cs.UTF_8
Serialization stack:
	- object not serializable (class: sun.nio.cs.UTF_8, value: UTF-8)
	- field (class: org.apache.spark.rdd.PipedRDD, name: org$apache$spark$rdd$PipedRDD$$encoding, type: class java.nio.charset.Charset)
	- object (class org.apache.spark.rdd.PipedRDD, PipedRDD[1] at pipe at <console>:29)
	- field (class: org.apache.spark.rdd.RDD$$anonfun$collect$1, name: $outer, type: class org.apache.spark.rdd.RDD)
	- object (class org.apache.spark.rdd.RDD$$anonfun$collect$1, <function0>)
	- field (class: org.apache.spark.rdd.RDD$$anonfun$collect$1$$anonfun$12, name: $outer, type: class org.apache.spark.rdd.RDD$$anonfun$collect$1)
	- object (class org.apache.spark.rdd.RDD$$anonfun$collect$1$$anonfun$12, <function1>)
	at org.apache.spark.serializer.SerializationDebugger$.improveException(SerializationDebugger.scala:40)
	at org.apache.spark.serializer.JavaSerializationStream.writeObject(JavaSerializer.scala:47)
	at org.apache.spark.serializer.JavaSerializerInstance.serialize(JavaSerializer.scala:101)
	at org.apache.spark.util.ClosureCleaner$.ensureSerializable(ClosureCleaner.scala:301)
	... 56 more
```

Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/rdd/RDD.scala]: > I will use String instead. Is that fine ?

👍

**Comment by tejasapatil** [comment]
ok to test

**Comment by tejasapatil** [comment]
Jenkins retest please

File: core/src/main/scala/org/apache/spark/rdd/PipedRDD.scala [hunk 6]
```diff
@@ -60,8 +61,7 @@ private[spark] class PipedRDD[T: ClassTag](
       printRDDElement: (T, String => Unit) => Unit = null,
       separateWorkingDir: Boolean = false) =
     this(prev, PipedRDD.tokenize(command), envVars, printPipeContext, printRDDElement,
-      separateWorkingDir, 8192)
-
+      separateWorkingDir, 8192, Codec.defaultCharsetCodec.name)
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/rdd/PipedRDD.scala]: I think we can remove this whole constructor actually. I don't see it is used anywhere and it's internal. It might help keep this a little simpler, if so. (The comment above it at least looks unrelated.)

**Comment by tejasapatil** [comment]
ok to test

**Comment by tejasapatil** [comment]
ok to test

**Comment by tejasapatil** [comment]
ok to test

**Comment by tejasapatil** [comment]
@zsxwing : I am done with changes from my end. Let me know if you have any more comments.

**Comment by zsxwing** [comment]
Could you update the PR title and description? Otherwise LGTM!

**Comment by tejasapatil** [comment]
@zsxwing : Updated the PR title, description and the Jira title. Thanks for your review !!

**Comment by zsxwing** [comment]
Thanks! Merging to master and 2.0.
